cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java

```diff
@@ -423,8 +427,13 @@ public ArtifactDetail addArtifact(Id.Artifact artifactId, File artifactFile,
                                     @Nullable Set<PluginClass> additionalPlugins,
                                     Map<String, String> properties) throws Exception {
     // check that the current user has write privileges on the namespace
-    authorizerInstantiatorService.get().enforce(artifactId.getNamespace().toEntityId(),
-                                                SecurityRequestContext.toPrincipal(), Action.WRITE);
+    // Skip authorization enforcement for system user. This is so the SystemArtifactLoader thread can add system
+    // artifacts.
+    Principal principal = SecurityRequestContext.toPrincipal();
+    NamespaceId namespace = artifactId.getNamespace().toEntityId();
+    if (!(Principal.SYSTEM.equals(principal) && NamespaceId.SYSTEM.equals(namespace))) {
```

Comment:
do we need similar checks for deletion? In case there are more auth related methods added in the future, perhaps it would be better to wrap the authorizerInstantiatorService with one that is a no-op if the system principal is being used?